ocamltest/run.c

```diff
@@ -0,0 +1,301 @@
+                                              : defaultLogger;
+  void *loggerData = settings->loggerData;
+  va_start(ap, msg);
+  mylog(logger, loggerData, "%s:%d: ", file, line);
+  logger(loggerData, msg, ap);
+  mylog(logger, loggerData, ": %s\n", strerror(errno));
+  va_end(ap);
+}
+
+#define myperror(msg, ...) \
+myperror_with_location(__FILE__, __LINE__, settings, msg, ## __VA_ARGS__)
+
+/* Same remark as for the error macro. */
+
+void open_error_with_location(
+  const char *file, int line,
+  const command_settings *settings,
+  const char *msg)
+{
+  myperror_with_location(file, line, settings, "Can not open %s", msg);
+}
+
+#define open_error(filename) \
+open_error_with_location(__FILE__, __LINE__, settings, filename)
+
+void realpath_error_with_location(
+  const char *file, int line,
+  const command_settings *settings,
+  const char *msg)
+{
+  myperror_with_location(file, line, settings, "realpath(\"%s\") failed", msg);
+}
+
+#define realpath_error(filename) \
+realpath_error_with_location(__FILE__, __LINE__, settings, filename)
+
+void handle_alarm(int sig)
+{
+  timeout_expired = 1;
+}
+
+int run_command_child(const command_settings *settings)
+{
+  int res;
+  int stdin_fd = -1, stdout_fd = -1, stderr_fd = -1; /* -1 means not redirected */
+  int inputFlags = O_RDONLY;
+  int outputFlags = O_CREAT | O_WRONLY | (settings->append ? O_APPEND : O_TRUNC);
+  int inputMode = 0400, outputMode = 0666;
+
+  if (setpgid(0, 0) == -1) myperror("setpgid");
+
+  if (is_defined(settings->stdin_filename))
+  {
+    stdin_fd = open(settings->stdin_filename, inputFlags, inputMode);
+    if (stdin_fd < 0)
+      open_error(settings->stdin_filename);
+    if ( dup2(stdin_fd, STDIN_FILENO) == -1 )
+      myperror("dup2 for stdin");
+  }
+
+  if (is_defined(settings->stdout_filename))
+  {
+    stdout_fd = open(settings->stdout_filename, outputFlags, outputMode);
+    if (stdout_fd < 0)
+      open_error(settings->stdout_filename);
+    if ( dup2(stdout_fd, STDOUT_FILENO) == -1 )
+      myperror("dup2 for stdout");
+  }
+
+  if (is_defined(settings->stderr_filename))
+  {
+    if (stdout_fd != -1)
+    {
+#ifdef __GLIBC__
+      char *stdout_realpath, *stderr_realpath;
+      stdout_realpath = realpath(settings->stdout_filename, NULL);
+      if (stdout_realpath == NULL)
+        realpath_error(settings->stdout_filename);
+      stderr_realpath = realpath(settings->stderr_filename, NULL);
+      if ( (stderr_realpath == NULL)  && (errno != ENOENT) )
+      {
+        free(stdout_realpath);
+        realpath_error(settings->stderr_filename);
+      }
+#else
+      char stdout_realpath[PATH_MAX], stderr_realpath[PATH_MAX];
+      if (realpath(settings->stdout_filename, stdout_realpath) == NULL)
+        realpath_error(settings->stdout_filename);
+      if ((realpath(settings->stderr_filename, stderr_realpath) == NULL) && (errno != ENOENT))
+        realpath_error(settings->stderr_filename);
+#endif /* __GLIBC__ */
```

Comment:
Xavier Leroy (2016/07/12 00:39 -0700):
> > +      stdout_realpath = realpath(settings->stdout_filename, NULL);
> > +      if (stdout_realpath == NULL)
> > +        realpath_error(settings->stdout_filename);
> > +      stderr_realpath = realpath(settings->stderr_filename, NULL);
> > +      if ( (stderr_realpath == NULL)  && (errno != ENOENT) )
> > +      {
> > +        free(stdout_realpath);
> > +        realpath_error(settings->stderr_filename);
> > +      }
> > +#else
> > +      char stdout_realpath[PATH_MAX], stderr_realpath[PATH_MAX];
> > +      if (realpath(settings->stdout_filename, stdout_realpath) == NULL)
> > +        realpath_error(settings->stdout_filename);
> > +      if ((realpath(settings->stderr_filename, stderr_realpath) == NULL) && (errno != ENOENT))
> > +        realpath_error(settings->stderr_filename);
> > +#endif /* __GLIBC__ */
> 
> It took me a while to understand what this code is doing.  I suggest moving it away in a function
> ```
> int paths_same_file(const char * path1, const char * path2)
> ```

Thanks. This has finally been done.
It turned out that the settings also needed to be propagated.